Tensile/Source/TensileTypes.h

```diff
@@ -77,106 +81,311 @@ TensileStatus tensileTeardown();
+    for (int i=0; i<NumSizes; i++) {
+      _sizes[i] = pdims.sizes(i);
+    }
+	}
+
+  bool operator< (const ProblemKey<NumSizes> & p) const
   {
     for (int i=0; i<NumSizes; i++) {
-      if (p.sizes[i] < this->sizes[i])
+      if (p._sizes[i] < this->_sizes[i])
         return true;
-      else if (p.sizes[i] > this->sizes[i])
+      else if (p._sizes[i] > this->_sizes[i])
         return false;
       // if equal, continue to next index
     }
     return false; // get here if all indices are equal
   };
 
-  SizeType lastSummationSize() const { return sizes[LastSummationIdx]; };
-  SizeType free0Size() const { return sizes[Free0Idx]; };
-  SizeType free1Size() const { return sizes[1]; };
+  bool operator== (const ProblemKey<NumSizes> & p) const
+  {
+    for (int i=0; i<NumSizes; i++) {
+      if (p._sizes[i] != this->_sizes[i])
+        return false;
+    }
+    return true;
+  };
+
+  size_t hash() const {
+    size_t h=0;
+    for (int i=0; i<NumSizes; i++) {
+      h ^= _sizes[i] + 0x9b9773e99e3779b9 + (h<<6) + (h>>2);
+    }
+    return h;
+  }
+
+  const SizeType sizes(int i) const { return _sizes[i];};
+  int numSizes() const { return NumSizes;};
+
+  std::ostream &print(std::ostream &os) const {
+    for (int i=0; i<NumSizes; i++) {
+      if (i != 0) {
+        os << ", ";
+      };
+      os << _sizes[i];
+    };
+    return os;
+  };
+
 
 private:
-  template<int I, typename T>
-  void init (T v) {
-    sizes[NumSizes-I] = v;
+  template<int I>
+  void init (SizeType v) {
+    _sizes[NumSizes-I-1] = v;
   }
 
-  template<int I, typename T, typename... Ts>
-  void init (T v, Ts... args )
-  {
-    sizes[NumSizes-I] = v;
+  template<int I, typename... Ts>
+  void init (SizeType v, Ts... args ) {
+    _sizes[NumSizes-I-1] = v;
     init<I-1> (args...);
   }
 
-
 private:
   // Data members:
-  SizeType sizes[NumSizes];
+  SizeType _sizes[NumSizes];
+};
+
+//-------------
+// Distance Functions between two problem sizes - used to find nearest neighbor or closest solution
+// Assumes p1 and p2 have same number of sizes
+//-------------
+
+template <class ProblemKeyType>
+struct RatioDistance {
+  double operator() (const ProblemKeyType &p1, const ProblemKeyType &p2) const
+  {
+    double distance = 1.0;
+    for (int i=0; i<p1.numSizes(); i++) {
+      distance += ::fabs(::log(double(p1.sizes(i))/double(p2.sizes(i))));
+    }
+    return distance;
+  }
+};
+
+template <class ProblemKeyType>
+struct ManhattanDistance {
+  double operator() (const ProblemKeyType &p1, const ProblemKeyType &p2) const
+  {
+    double distance = 0;
+    for (int i=0; i<p1.numSizes(); i++) {
+      distance += ::fabs(double(p1.sizes(i)) - double(p2.sizes(i)));
+    }
+    return distance;
+  }
+};
+
+
+template <class ProblemKeyType>
+struct EuclideanDistance {
+  double operator() (const ProblemKeyType &p1, const ProblemKeyType &p2) const
+  {
+    double distance = 0;
+    for (int i=0; i<p1.numSizes(); i++) {
+      if (p1.sizes(i) > p2.sizes(i))
```

Comment:
can this be speeded up with:
 
double t = p1.sizes(i) - p2.sizes(i);
distance += t * t;

Reply:
sizes are unsigned and in some case p2 may be bigger than p1 - subtracting would create large unsigned number.  This code is not often on critical path.